packages/clerk-js/src/utils/sha1.ts

```diff
@@ -0,0 +1,145 @@
+/////////
+///////
+// The SHA (Secure Hash Algorithm) can be used to encrypt data for secure transfer between applications.
+// The SHA1() function returns a string with the SHA1 encrypted hash as a string with 40 characters.
+// It is fully compatible with UTF-8 encoding.
+///
+//
+export const SHA1 = (msg: string) => {
```

Comment:
Team, do you know if this polyfill is necessary? We could use webcrypto. It was great browser support.

```async function sha1(str) {
  const enc = new TextEncoder();
  const hash = await crypto.subtle.digest('SHA-1', enc.encode(str));
  return Array.from(new Uint8Array(hash))
    .map(v => v.toString(16).padStart(2, '0'))
    .join('');
}```

Reply:
@SokratisVidros unfortunately the implementation you provided is async. Going this route means we will have to make `authStateKey` a function(similar to getToken) and let the user resolve it. I don't think we want this. This was the main reason we didn't choose this kind of handling and went with a more straight forward sync way.

cc @nikosdouvlis

Reply:
This was discussed in this PR again, the initial design described a sync `authStateKey` prop. However, given that this has come up twice so far, let's explore what making this async would look like.

```
import useSWR from 'swr';
import { useAuth } from '@clerk/nextjs';

export default function useClerkSWR(url) {
  const { getToken, authStateKey } = useAuth();

  const fetcher = async (...args) => {
    return fetch(...args, {
     // this is valid as this Promise is correctly handled within SWR 
      headers: { Authorization: `Bearer ${await getToken()}` }
    }).then(res => res.json());
  };

  // await cannot be used here currently
  return useSWR([url, authStateKey], fetcher);
}
```
(References: https://swr.vercel.app/docs/arguments)

Meaning that if we were to use the built-in crypto methods, authStateKey would need to:
1. either be async (bad DX as it requires manually unwrapping the value from the promise inside hooks)
2. or, be a sync variable holding `undefined` or `""` while the Promise is not resolved, and a `string` once it resolves. If this is implemented with `useState`, `useAuth` will render multiple times.

A different approach would be to use `useRef` within `useAuth` (so no extra renders because of `useState) to hold a reference to the latest resolved key and expose a sync method to either retrieve the current value of the promise if the promise is resolved, or fallback to a different value (eg `""`) if the promise is not resolved

Reply:
I agree with @SokratisVidros. Let's use webcrypto. It should work since localhost is considered a secure context.

Reply:
@desiprisg Please take a look at my answer above as this possible but not as straightforward as we'd like

Reply:
Great call on it being async.